lib/iris/fileformats/grib/message.py

```diff
@@ -52,31 +66,46 @@ def messages_from_filename(filename):
             Name of the file to generate fields from.
 
         """
-        with open(filename, 'rb') as grib_fh:
-            while True:
-                offset = grib_fh.tell()
-                grib_id = gribapi.grib_new_from_file(grib_fh)
-                if grib_id is None:
-                    break
-                raw_message = _RawGribMessage(grib_id)
-                recreate_raw = _MessageLocation(filename, offset)
-                yield _GribMessage(raw_message, recreate_raw)
-
-    def __init__(self, raw_message, recreate_raw):
+        grib_fh = open(filename, 'rb')
+        file_ref = _OpenFileRef(grib_fh)
+
+        while True:
+            offset = grib_fh.tell()
+            grib_id = gribapi.grib_new_from_file(grib_fh)
+            if grib_id is None:
+                break
+            raw_message = _RawGribMessage(grib_id)
+            recreate_raw = _MessageLocation(filename, offset)
+            yield GribMessage(raw_message, recreate_raw, file_ref=file_ref)
+
+    def __init__(self, raw_message, recreate_raw, file_ref=None):
         """
-
-        Args:
-
-        * raw_message:
-            The _RawGribMessage instance which should be wrapped to
-            provide the `data` attribute.
+        It is recommended to obtain GribMessage instance from the static method
+        :meth:`GribMessage.messages_from_filename`, rather than creating
+        them directly.
 
         """
         self._raw_message = raw_message
```

Comment:
Some (private) comments here clarifying the purpose of the constructor args would be good.
Maybe something like... :
```
    # A RawGribMessage giving gribapi access to the original grib message.
    self._raw_message = raw_message
    # A _MessageLocation which biggus uses to read the message data array,
    # by which time this message may be dead and the original grib file closed.
    self._recreate_raw = recreate_raw
    # An _OpenFileRef to keep the grib file open while this GribMessage is alive,
    # so that we can always use self._raw_message to fetch keys.
    self._file_ref = file_ref
```

I admit that seems a bit wordy though...